daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -483,41 +481,82 @@ private[validation] object Typing {
       AlphaEquiv.alphaEquiv(t1, t2) ||
         AlphaEquiv.alphaEquiv(expandTypeSynonyms(t1), expandTypeSynonyms(t2))
 
-    private def checkIfaceImplementations(
+    private def checkGenImplementation(
         tplTcon: TypeConName,
-        impls: Map[TypeConName, TemplateImplements],
+        ifaceTcon: TypeConName,
+        implMethods: List[(MethodName, Expr)],
     ): Unit = {
+      val DefInterfaceSignature(requires, _, _, methods, _, _) =
+        handleLookup(ctx, interface.lookupInterface(ifaceTcon))
 
-      impls.foreach { case (iface, impl) =>
-        val DefInterfaceSignature(requires, _, _, methods, _, _) =
-          handleLookup(ctx, interface.lookupInterface(impl.interfaceId))
-
-        requires
-          .filterNot(impls.contains)
-          .foreach(required => throw EMissingRequiredInterface(ctx, tplTcon, iface, required))
+      requires
+        .filterNot(required =>
+          interface.lookupInterfaceCoImplements(tplTcon, required).isRight ||
```

Comment:
side note: we should really rename `interface` here. This is super confusing.t

Reply:
Please take care of that in a special PR.